File: java/src/main/java/org/solutions/leetcode/medium/RedundantConnection.java
package org.solutions.leetcode.medium;

import org.model.DSU;

import java.util.ArrayList;
import java.util.List;

/**
 * <a href="https://leetcode.com/problems/redundant-connection">684. Redundant Connection</a>
 */

@SuppressWarnings("SuspiciousNameCombination")
public class RedundantConnection {
    public int[] findRedundantConnection(int[][] edges) {
        int n = 0;

        for (int[] edge : edges) {
            n = Math.max(n, Math.max(edge[0], edge[1]));
        }

        var dsu = new DSU(n + 1);
        List<int[]> result = new ArrayList<>();

        for (int[] edge : edges) {
            int x = edge[0];
            int y = edge[1];

            if (dsu.find(x) == dsu.find(y)) {
                result.add(edge);
            } else {
                dsu.union(x, y);
            }
        }

        return result.getLast();
    }
}

File: golang/solution/leetcode/medium/redundant_connection.go
package medium

import "golang/model"

// https://leetcode.com/problems/redundant-connection 684. Redundant Connection

func findRedundantConnection(edges [][]int) []int {
	n := 0
	for _, edge := range edges {
		n = max(n, max(edge[0], edge[1]))
	}

	dsu := model.NewDSU(1, n+1)
	var res [][]int

	for _, edge := range edges {
		x, y := edge[0], edge[1]
		if dsu.Find(x) == dsu.Find(y) {
			res = append(res, edge)
		} else {
			dsu.Union(x, y)
		}
	}

	return res[len(res)-1]
}

func FindRedundantConnection(edges [][]int) []int {
	return findRedundantConnection(edges)
}
